extensions/Worker.Extensions.CosmosDB/src/CosmosDBConverter.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.Azure.Functions.Worker.Core;
+using Azure.Cosmos;
+using System.Reflection;
+
+namespace Microsoft.Azure.Functions.Worker.Converters
+{
+    /// <summary>
+    /// Converter to bind Blob Storage type parameters.
+    /// </summary>
+    internal class CosmosDBConverter : IInputConverter
+    {
+        public ValueTask<ConversionResult> ConvertAsync(ConverterContext context)
+        {
+            if (context.Source is not IBindingData bindingData)
+            {
+                return new ValueTask<ConversionResult>(ConversionResult.Unhandled());
+            }
+
+            if (bindingData.Source is not "Microsoft.Azure.WebJobs.Extensions.CosmosDB")
+            {
+                return new ValueTask<ConversionResult>(ConversionResult.Unhandled());
+            }
+
+            try
+            {
+                object result = ToTargetType(context.TargetType, bindingData.Content);
+
+                if (result is not null)
+                {
+                    return new ValueTask<ConversionResult>(ConversionResult.Success(result));
+                }
+            }
+            catch (Exception ex)
+            {
+                // TODO: DeserializeObject could throw
+                Console.WriteLine(ex);
+            }
+
+            return new ValueTask<ConversionResult>(ConversionResult.Unhandled());
+        }
+
+        private object? ToTargetType(Type targetType, BinaryData content) => targetType switch
+        {
+            Type _ when targetType == typeof(CosmosClient)              => CreateCosmosReference<CosmosClient>(content),
+            Type _ when targetType == typeof(CosmosDatabase)            => CreateCosmosReference<CosmosDatabase>(content),
+            Type _ when targetType == typeof(CosmosContainer)           => CreateCosmosReference<CosmosContainer>(content),
+            _ => CreateTargetObject(targetType, content)
+        };
+
+        private object CreateTargetObject(Type targetType, BinaryData content)
+        {
+            MethodInfo deserializeObjectMethod = GetType()
+                                                .GetMethod(nameof(DeserializeTargetObject), BindingFlags.Instance | BindingFlags.NonPublic)
+                                                .MakeGenericMethod(new Type[] { targetType });
+            return deserializeObjectMethod.Invoke(this, new object[] { content });
```

Comment:
Interesting `BinaryData` does not expose this itself, but we can roll our own extension method to avoid the reflection here.

Here is the `ToObjectFromJson<T>` implementation of BinaryData. It already is just using `typeof(T)` so it could so easily have a non-generic overload which takes in `Type type`.

``` CSharp
    public T ToObjectFromJson<T>(JsonSerializerOptions? options = null)
    {
        return (T)JsonSerializer.Deserialize(_bytes.Span, typeof(T), options);
    }
```

We can add:

``` CSharp
    public static object? ToObjectFromJson(this BinaryData data, Type type, JsonSerializerOptions? options = null)
    {
        return JsonSerializer.Deserialize(_bytes.Span, type, options);
    }
```

Reply:
Was trying to figure out how to avoid reflection so definitely appreciate the suggestion here - I'll try it out!